### ISSUE TITLE
Interrupts are not processed in CreateElementPages for HNSW index

### ISSUE THREAD

**Issue by knizhnik** [pgvector/pgvector#202]
There is CHECK_FOR_INTERRUPTS() in the loop in `CreateElementPages` but actually processing of interrupts is hold  because locking buffer  increments `InterruptHoldoffCount`. And in this loop some buffer is always locked.
Proposed patch: 
```
diff --git a/src/hnswbuild.c b/src/hnswbuild.c
index 90ff9c3..0f68726 100644
--- a/src/hnswbuild.c
+++ b/src/hnswbuild.c
@@ -126,9 +126,11 @@ CreateElementPages(HnswBuildState * buildstate)
                {
                        /* Add a new page */
                        Buffer          newbuf = HnswNewBuffer(index, forkNum);
+                       BlockNumber     newblkno = BufferGetBlockNumber(newbuf);
 
                        /* Update previous page */
-                       HnswPageGetOpaque(page)->nextblkno = BufferGetBlockNumber(newbuf);
+                       HnswPageGetOpaque(page)->nextblkno = newblkno;
+                       LockBuffer(newbuf, BUFFER_LOCK_UNLOCK);
 
                        /* Commit */
                        MarkBufferDirty(buf);
@@ -141,6 +143,7 @@ CreateElementPages(HnswBuildState * buildstate)
 
                        /* Prepare new page */
                        buf = newbuf;
+                       LockBuffer(buf, BUFFER_LOCK_EXCLUSIVE);
                        state = GenericXLogStart(index);
                        page = GenericXLogRegisterBuffer(state, buf, GENERIC_XLOG_FULL_IMAGE);
                        HnswInitPage(buf, page);
```

**Comment by ankane** [comment]
Thanks @knizhnik, nice find. I'm currently working on placing elements and neighbors on the same page when possible (which affects this code), but will make sure this gets fixed before the release.

**Comment by ankane** [comment]
Fixed in 3169a60e5cc7b7c9f6866362b990c74fc49af0bd. Thanks again for reporting.
